setup.py

```diff
@@ -177,7 +188,109 @@ def build_extension(self, ext):
+                        NANOARROW_ARROW_ITERATOR_SRC_DIR, "Python", "Helpers.cpp"
+                    ),
+                    os.path.join(NANOARROW_ARROW_ITERATOR_SRC_DIR, "Util", "time.cpp"),
+                    NANOARROW_LOGGING_SRC_DIR + "/logging.cpp",
+                    os.path.join(NANOARROW_ARROW_ITERATOR_SRC_DIR, "nanoarrow.c"),
+                    os.path.join(NANOARROW_ARROW_ITERATOR_SRC_DIR, "nanoarrow_ipc.c"),
+                    os.path.join(NANOARROW_ARROW_ITERATOR_SRC_DIR, "flatcc.c"),
+                ]
+                ext.include_dirs.append(NANOARROW_ARROW_ITERATOR_SRC_DIR)
+                ext.include_dirs.append(NANOARROW_LOGGING_SRC_DIR)
+
+                if sys.platform == "win32":
+                    if not any("/std" not in s for s in ext.extra_compile_args):
+                        ext.extra_compile_args.append("/std:c++17")
+                elif sys.platform == "linux" or sys.platform == "darwin":
+                    if "std=" not in os.environ.get("CXXFLAGS", ""):
+                        ext.extra_compile_args.append("-std=c++17")
+                        ext.extra_compile_args.append("-D_GLIBCXX_USE_CXX11_ABI=0")
+                    if sys.platform == "darwin":
+                        ext.extra_compile_args.append("-mmacosx-version-min=10.13")
+
+                ext.library_dirs.append(
+                    os.path.join(current_dir, self.build_lib, "snowflake", "connector")
+                )
+
+                # sys.platform for linux used to return with version suffix, (i.e. linux2, linux3)
+                # After version 3.3, it will always be just 'linux'
+                # https://docs.python.org/3/library/sys.html#sys.platform
+                if sys.platform == "linux":
+                    ext.extra_link_args += ["-Wl,-rpath,$ORIGIN"]
+                elif sys.platform == "darwin":
+                    # rpath,$ORIGIN only work on linux, did not work on darwin. use @loader_path instead
+                    # fyi, https://medium.com/@donblas/fun-with-rpath-otool-and-install-name-tool-e3e41ae86172
+                    ext.extra_link_args += ["-rpath", "@loader_path"]
+
+            original__compile = self.compiler._compile
+
+            # the following is required by nanoarrow to compile c files
+            def new__compile(obj, src: str, ext, cc_args, extra_postargs, pp_opts):
+                if (
+                    src.endswith("nanoarrow.c")
+                    or src.endswith("nanoarrow_ipc.c")
+                    or src.endswith("flatcc.c")
+                ):
+                    extra_postargs = [s for s in extra_postargs if s != "-std=c++17"]
+                return original__compile(
+                    obj, src, ext, cc_args, extra_postargs, pp_opts
+                )
+
+            self.compiler._compile = new__compile
+
+            try:
+                build_ext.build_extension(self, ext)
+            finally:
+                del self.compiler._compile
```

Comment:
Question: what happens when the extension fails to build?

Reply:
if build fails, it will lead to installation failure

Reply:
resolved #1758